test/reputation-system/dispute-resolution-misbehaviour.js

```diff
@@ -338,18 +331,230 @@ contract("Reputation Mining - disputes resolution misbehaviour", accounts => {
+      }
+
+      await forwardTime(MINING_CYCLE_DURATION / 2, this);
+
+      console.log("Starting disputes");
+
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[0], clients[1]);
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[2], clients[3]);
+      await accommodateChallengeAndInvalidateHash(colonyNetwork, this, clients[8]);
+
+      await checkErrorRevert(
+        accommodateChallengeAndInvalidateHash(colonyNetwork, this, clients[8]),
+        "colony-reputation-mining-previous-dispute-round-not-complete"
+      );
+
+      console.log("Cleaning up");
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[4], clients[5]);
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[6], clients[7]);
+
+      // Round 1 finished. Move to round 2
+
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[0], clients[2]);
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[8], clients[4]);
+      await accommodateChallengeAndInvalidateHash(colonyNetwork, this, clients[6]);
+
+      // Round 2 finished.
+
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[0], clients[8]);
+      await accommodateChallengeAndInvalidateHash(colonyNetwork, this, clients[6]);
+      await accommodateChallengeAndInvalidateHashViaTimeout(colonyNetwork, this, clients[0], clients[6]);
+      await repCycle.confirmNewHash(4);
+    });
+
+    it(`should prevent a hash from advancing if it might still get an opponent,
+     even if that opponent is from more than one round ago`, async function advancingTest() {
+      this.timeout(10000000);
+
+      expect(accounts.length, "Not enough accounts for test to run").to.be.at.least(17);
+      const accountsForTest = accounts.slice(3, 17);
+
+      await fundColonyWithTokens(metaColony, clnyToken, INITIAL_FUNDING.muln(14));
+      for (let i = 0; i < 14; i += 1) {
+        await giveUserCLNYTokensAndStake(colonyNetwork, accountsForTest[i], DEFAULT_STAKE);
+        await setupFinalizedTask({ colonyNetwork, colony: metaColony, worker: accountsForTest[i] });
+        // These have to be done sequentially because this function uses the total number of tasks as a proxy for getting the
+        // right taskId, so if they're all created at once it messes up.
+      }
+
+      // We need to complete the current reputation cycle so that all the required log entries are present
+      await advanceMiningCycleNoContest({ colonyNetwork, test: this });
+
+      const clients = await Promise.all(
+        accountsForTest.map(async (addr, index) => {
+          const client = new MaliciousReputationMinerExtraRep(
+            { loader, realProviderPort, useJsTree, minerAddress: addr },
+            accountsForTest.length - index,
+            index
+          );
```

Comment:
I wonder if there's a possible helper function which handles this setup logic -- I'm seeing a lot of 1) create X tasks, 2) create X miners, 3) submit X hashes.